alacritty_terminal/src/tty/windows/blocking.rs

```diff
@@ -0,0 +1,285 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+//! Code for running a reader/writer on another thread while driving it through `polling`.
+
+use piper::{pipe, Reader, Writer};
+use polling::os::iocp::{CompletionPacket, PollerIocpExt};
+use polling::{Event, PollMode, Poller};
+
+use std::io::prelude::*;
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Wake, Waker};
+use std::{io, thread};
+
+struct Registration {
+    interest: Mutex<Option<Interest>>,
+    end: End,
+}
+
+#[derive(Copy, Clone)]
+enum End {
+    Reader,
+    Writer,
+}
+
+struct Interest {
+    /// The event to send about completion.
+    event: Event,
+
+    /// The poller to send the event to.
+    poller: Arc<Poller>,
+
+    /// The mode that we are in.
+    mode: PollMode,
+}
+
+/// Poll a reader in another thread.
+pub struct UnblockedReader<R> {
+    /// The event to send about completion.
+    interest: Arc<Registration>,
+
+    /// The pipe that we are reading from.
+    pipe: Reader,
+
+    /// Is this the first time registering?
+    first_register: bool,
+
+    /// We logically own the reader, but we don't actually use it.
+    _reader: PhantomData<R>,
+}
+
+impl<R: Read + Send + 'static> UnblockedReader<R> {
+    /// Spawn a new unblocked reader.
+    pub fn new(mut source: R, pipe_capacity: usize) -> Self {
+        // Create a new pipe.
+        let (reader, mut writer) = pipe(pipe_capacity);
+        let interest = Arc::new(Registration {
+            interest: Mutex::<Option<Interest>>::new(None),
+            end: End::Reader,
+        });
+
+        // Spawn the reader thread.
+        thread::Builder::new()
+            .name("alacritty-tty-reader-thread".into())
```

Comment:
Pretty sure we have a utility to spawn a named thread that should make this a little less verbose.